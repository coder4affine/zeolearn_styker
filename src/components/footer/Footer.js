import React from 'react';
import PropTypes from 'prop-types';
import Button from '../button/Button';

const Footer = ({ onSelect, select }) => {
  return (
    <div style={{ display: 'flex', flexDirection: 'row', height: 50 }}>
      <Button
        style={{ flex: 1, backgroundColor: select === 'all' ? 'green' : 'red' }}
        onClick={() => onSelect('all')}
      >
        <span>All</span>
      </Button>
      <Button
        style={{ flex: 1, backgroundColor: select === 'pending' ? 'green' : 'red' }}
        onClick={() => onSelect('pending')}
      >
        <span>Pending</span>
      </Button>
      <Button
        style={{ flex: 1, backgroundColor: select === 'completed' ? 'green' : 'red' }}
        onClick={() => onSelect('completed')}
      >
        <span>Completed</span>
      </Button>
    </div>
  );
};

Footer.propTypes = {
  onSelect: PropTypes.func.isRequired,
  select: PropTypes.string.isRequired,
};

export default Footer;
