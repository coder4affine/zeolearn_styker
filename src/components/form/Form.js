/* eslint-disable react/static-property-placement */
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import TextInput from '../textInput/TextInput';
import Button from '../button/Button';
import styles from './styles';

export default class Form extends PureComponent {
  text = '';

  static propTypes = {
    onAddTodo: PropTypes.func.isRequired,
  };

  addTodo = e => {
    e.preventDefault();
    const { onAddTodo } = this.props;
    onAddTodo(this.textInput.value);
    this.textInput.value = '';
  };

  render() {
    return (
      <form onSubmit={this.addTodo}>
        <div style={styles.container}>
          <TextInput
            forwardRed={ref => {
              this.textInput = ref;
            }}
            type="text"
            name="text"
            style={{
              height: 30,
            }}
            placeholder="Please Write your TODO Here"
            required
          />
          <Button type="submit">
            <span>Add Todo</span>
          </Button>
        </div>
      </form>
    );
  }
}
