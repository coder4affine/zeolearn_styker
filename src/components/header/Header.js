import React from 'react';
import PropTypes from 'prop-types';
import { LocaleConsumer } from '../../context/localeContext';
import styles from './style';

const Header = ({ title }) => {
  console.log('Header');
  return (
    <div style={styles.container}>
      <h3>{title}</h3>
      <LocaleConsumer>
        {value => {
          console.log('value');
          return (
            <div>
              <h1>{value.locale}</h1>
              <button type="button" onClick={() => value.changeLocale('german')}>
                Change Locale
              </button>
            </div>
          );
        }}
      </LocaleConsumer>
    </div>
  );
};

Header.propTypes = {
  title: PropTypes.string.isRequired,
};

export default Header;
