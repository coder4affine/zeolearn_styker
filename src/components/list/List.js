import React, { memo } from 'react';
import PropTypes from 'prop-types';
import Button from '../button/Button';

const List = ({ todos, onToggleComplete, onDeleteTodo }) => {
  console.log('List');
  return (
    <>
      {todos.map(todo => (
        <div key={todo.id} style={{ display: 'flex', flexDirection: 'row', padding: 10 }}>
          <input type="checkbox" checked={todo.isDone} onChange={() => onToggleComplete(todo)} />
          <span style={{ display: 'flex', flex: 1, padding: '0 10px' }}>{todo.text}</span>
          <Button onClick={() => onDeleteTodo(todo)}>
            <span>Delete Todo</span>
          </Button>
        </div>
      ))}
    </>
  );
};

List.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
      isDone: PropTypes.bool.isRequired,
    }),
  ).isRequired,
  onToggleComplete: PropTypes.func.isRequired,
  onDeleteTodo: PropTypes.func.isRequired,
};

export default memo(List);
