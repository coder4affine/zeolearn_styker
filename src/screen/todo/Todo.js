import React, { useState, useEffect } from 'react';
import Header from '../../components/header/Header';
import Form from '../../components/form/Form';
import List from '../../components/list/List';
import Footer from '../../components/footer/Footer';
import './todo.css';

const Todo = () => {
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');
  const [todos, setTodos] = useState([]);
  const [select, setSelect] = useState('all');

  useEffect(() => {
    const loadData = async () => {
      setLoading(true);
      try {
        const res = await fetch('http://localhost:3004/todos');
        const newTodos = await res.json();
        setTodos(newTodos);
      } catch (err) {
        setError(err.message);
      } finally {
        setLoading(false);
      }
    };

    const copy = () => {
      console.log('copied');
    };

    loadData();
    document.addEventListener('copy', copy);

    return () => {
      document.removeEventListener('copy', copy);
    };
  }, []);

  useEffect(() => {
    console.log('useEffect 1');
  }, [loading, error, select, todos]);

  const onToggleComplete = async todo => {
    setLoading(true);
    try {
      const updatedTodo = {
        ...todo,
        isDone: !todo.isDone,
      };

      await new Promise(resolve => {
        setTimeout(() => {
          resolve();
        }, 2000);
      });

      const res = await fetch(`http://localhost:3004/todos/${todo.id}`, {
        method: 'PUT',
        body: JSON.stringify(updatedTodo),
        headers: {
          accept: 'application/json',
          'Content-Type': 'application/json',
        },
      });

      const newUpdatedTodo = await res.json();

      const index = todos.findIndex(x => x.id === todo.id);
      setTodos([...todos.slice(0, index), newUpdatedTodo, ...todos.slice(index + 1)]);
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };

  const onDeleteTodo = async todo => {
    setLoading(true);
    try {
      await fetch(`http://localhost:3004/todos/${todo.id}`, {
        method: 'DELETE',
      });

      const updatedTodos = todos.filter(x => x.id !== todo.id);
      setTodos(updatedTodos);
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };

  const onSelect = val => {
    setSelect(val);
  };

  const onAddTodo = async todo => {
    setLoading(true);
    try {
      const res = await fetch('http://localhost:3004/todos', {
        method: 'POST',
        body: JSON.stringify({ text: todo, isDone: false }),
        headers: {
          accept: 'application/json',
          'Content-Type': 'application/json',
        },
      });

      const newTodo = await res.json();
      setTodos([...todos, newTodo]);
    } catch (err) {
      setError(err.message);
    } finally {
      setLoading(false);
    }
  };

  const filteredTodos = () => {
    return todos.filter(x => {
      switch (select) {
        case 'pending':
          return !x.isDone;
        case 'completed':
          return x.isDone;
        default:
          return true;
      }
    });
  };

  return (
    <div
      id="todo"
      style={{
        flex: 1,
        display: 'flex',
        flexDirection: 'column',
        height: '100vh',
      }}
    >
      {/* implement Header */}
      <Header title="Todo App" />
      {loading && <h1 style={{ textAlign: 'center' }}>Loading....</h1>}
      {error && <h1 style={{ textAlign: 'center' }}>{error}</h1>}
      <Form onAddTodo={onAddTodo} />
      <div style={{ flex: 1 }}>
        <List
          todos={filteredTodos()}
          onToggleComplete={onToggleComplete}
          onDeleteTodo={onDeleteTodo}
        />
      </div>
      <Footer onSelect={onSelect} select={select} />
    </div>
  );
};

export default Todo;

// export default class Todo extends Component {
//   state = {
//     todos: [],
//     select: 'all',
//     loading: false,
//     error: '',
//   };

//   async componentDidMount() {
//     this.setState({ loading: true });
//     try {
//       const res = await fetch('http://localhost:3004/todos');
//       const todos = await res.json();
//       this.setState({ todos });
//     } catch (error) {
//       this.setState({ error: error.message });
//     } finally {
//       this.setState({ loading: false });
//     }
//   }

//   render() {
//     const { todos, select, loading, error } = this.state;

// ;
//   }
// }
