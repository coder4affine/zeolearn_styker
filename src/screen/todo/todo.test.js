import React from 'react';
import { shallow } from 'enzyme';
import Header from '../../components/header/Header';
import Form from '../../components/form/Form';
import List from '../../components/list/List';
import Footer from '../../components/footer/Footer';

// Todo
import Todo from './Todo';

function setup() {
  const props = {};
  const wrapper = shallow(<Todo />);
  return { wrapper, props };
}

describe('Snapshot of Todo Component', () => {
  it('Snapshot', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});

describe('WelcomeMessage Test Suite', () => {
  it('Should have an image', () => {
    const { wrapper } = setup();
    expect(wrapper.find('div').exists()).toBe(true);
  });
});

describe('Check Header ', () => {
  it('Check header exist of not', () => {
    const { wrapper } = setup();
    expect(wrapper.find(Header).exists()).toBe(true);
  });

  it('check title Exist in header component', () => {
    const { wrapper } = setup();
    expect(wrapper.find(Header).props()).toEqual({ title: 'Todo App' });
  });
});

describe('Check Form ', () => {
  it('Check Form exist of not', () => {
    const { wrapper } = setup();
    expect(wrapper.find(Form).exists()).toBe(true);
  });

  it('Check Form exist of not', () => {
    const { wrapper } = setup();
    expect(wrapper.find(Form).props().onAddTodo).toEqual(expect.any(Function));
  });
});

it('Check List exist of not', () => {
  const { wrapper } = setup();
  expect(wrapper.find(List).exists()).toBe(true);
});

it('Check Footer exist of not', () => {
  const { wrapper } = setup();
  expect(wrapper.find(Footer).exists()).toBe(true);
});
