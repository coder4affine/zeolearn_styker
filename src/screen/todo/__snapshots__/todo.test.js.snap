// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot of Todo Component Snapshot 1`] = `
<div
  id="todo"
  style={
    Object {
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "height": "100vh",
    }
  }
>
  <Header
    title="Todo App"
  />
  <Form
    onAddTodo={[Function]}
  />
  <div
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <List
      onDeleteTodo={[Function]}
      onToggleComplete={[Function]}
      todos={Array []}
    />
  </div>
  <Footer
    onSelect={[Function]}
    select="all"
  />
</div>
`;
